### PR TITLE
Fix find_first_block seek logic

### PR DESCRIPTION
**Pull request by nyanzebra**
From looking at the backing file for RingBuffer from LH tests, the reason why there was an infinite loop of erroring from RingBuffer is due to bad seek logic for finding the first block.

This should fix the problem.

### CODE REVIEW

**Comment by nyanzebra** [comment]
Will also cherry-pick into 1.2 and 1.1 once this is approved.

**Comment by nyanzebra** [comment]
Just to make things even clearer, before we were seeking with SeekFrom::Current and now we are using SeekFrom::Start. We want to not skip over the bytes we have read already in the buffer by using SeekFrom::Current as it means we might throw away a perfectly good block. Note read will update the buffer pointer! Hence Current is wrong :)

File: mqtt/mqtt-bridge/src/persist/waking_state/ring_buffer/mod.rs
Comment (varunpuranik) [inline mqtt/mqtt-bridge/src/persist/waking_state/ring_buffer/mod.rs:516]: Why do we need this change in logic? What was the bug exactly?

Also, tests?

**Comment by varunpuranik** [comment]
> Will also cherry-pick into 1.2 and 1.1 once this is approved.

Please don't cherry pick changes to any of the release branches without shiproom approval...

File: mqtt/mqtt-bridge/src/persist/waking_state/ring_buffer/mod.rs
Comment (nyanzebra) [reply mqtt/mqtt-bridge/src/persist/waking_state/ring_buffer/mod.rs:516]: Bug is mainly using SeekFrom::Current instead of SeekFrom::Start. When we do something like this:

```rust
let mut buf = vec![];
file.read(&mut buf)?;
```
The seek pointer will move to the amount of bytes read. So if we did not find the block in the buffer then seek with current what happens will look like this:
Pretend following is true:
File size: 100
Buf size: 40
+------------------------+
| File data                         |
+------------------------+
 \        \  <---\----- SeekPtr is 40 now after read
  +-----+       |
   | buf   |        |
  +-----+       |
       \             |
        +-------+
         | block   | <- where block is really, say from 10-50 of the file
        +-------+

If we continue to use current then we will have read 0-40 then 40-80, thus missing the first block in the file. That is why we need to do something like SeekFrom::Start, so we instead do 0-40, 1-41, 2-42, etc...

Comment (nyanzebra) [reply mqtt/mqtt-bridge/src/persist/waking_state/ring_buffer/mod.rs:516]: test added

Comment (varunpuranik) [reply mqtt/mqtt-bridge/src/persist/waking_state/ring_buffer/mod.rs:516]: I am still a bit confused.. let's chat. 
But meanwhile, can you pls say what you said above to the PR as comments?

File: builds/checkin/mqtt.yaml
Comment (nyanzebra) [inline builds/checkin/mqtt.yaml:48]: Added to speed up tests (should build and run release) if this improves time to run and stops checkin from getting cancelled then I will open a new pr with just this change

Comment (nyanzebra) [reply builds/checkin/mqtt.yaml:48]: Seems to have improved time significantly! Created a pr for this small change, and if that goes in then can rebase this pr

File: mqtt/mqtt-bridge/src/persist/waking_state/ring_buffer/mod.rs
Comment (and-rewsmith) [inline mqtt/mqtt-bridge/src/persist/waking_state/ring_buffer/mod.rs:553]: Makes sense :+1:. If this can never find a block then `read_exact()` will error eventually right?

File: mqtt/mqtt-bridge/src/persist/waking_state/ring_buffer/mod.rs
Comment (and-rewsmith) [inline mqtt/mqtt-bridge/src/persist/waking_state/ring_buffer/mod.rs:419]: Can you give some context as to why this is changing?

**Review by and-rewsmith — COMMENTED** [review]
Synced with Robert. I agree with the high level approach just a couple questions.

File: mqtt/mqtt-bridge/src/persist/waking_state/ring_buffer/mod.rs
Comment (nyanzebra) [reply mqtt/mqtt-bridge/src/persist/waking_state/ring_buffer/mod.rs:553]: Yep, that is expected. In the `find_pointers_and_order_post_crash` fn, there is a check for any Err. If there is one, we return a default RingBuffer.

File: mqtt/mqtt-bridge/src/persist/waking_state/ring_buffer/mod.rs
Comment (nyanzebra) [reply mqtt/mqtt-bridge/src/persist/waking_state/ring_buffer/mod.rs:419]: Yeah, we already found the first block at this point, which has a write_index (this is the position of the write ptr)
So either:
A) the write ptr is at 0
B) the write ptr is at an offset away from 0

To elaborate further, we know that the read and write ptr must start at the beginning of a block. We have found the first block in the RB as we had scanned from 0->X to find it. If X is greater than 0, then it means all the bytes before X are likely data or an incomplete block, therefore we know starting at 0 is useless. That is why these now start at `start` instead of `0`, so that these begin at the first block we found and not potentially in middle of garbled bytes that mean nothing.

Sorry if this is still a bit complicated... Really looking forward to when I replace all of this with a simpler design :)

**Review by and-rewsmith — APPROVED** [review]

File: mqtt/mqtt-bridge/src/persist/waking_state/ring_buffer/mod.rs
Comment (and-rewsmith) [reply mqtt/mqtt-bridge/src/persist/waking_state/ring_buffer/mod.rs:419]: > We have found the first block in the RB as we had scanned from 0->X to find it. If X is greater than 0, then it means all the bytes before X are likely data or an incomplete block

If there is data before X, then shouldn't the read pointer be positioned before the write pointer so that we can read this data? I was expecting this to re-establish these pointers at specific locations, not necessarily the beginning of the block.

Comment (nyanzebra) [reply mqtt/mqtt-bridge/src/persist/waking_state/ring_buffer/mod.rs:419]: Ptrs have to start at a block is what I mean, perhaps I should've worded it as "... bytes before X are not a block"

File: mqtt/mqtt-bridge/src/persist/waking_state/ring_buffer/mod.rs
Comment (vadim-kovalyov) [inline mqtt/mqtt-bridge/src/persist/waking_state/ring_buffer/mod.rs:534]: just to confirm, our blocks are not aligned, so you need to check every single byte (and not 0,4,8,12,16, for example)?

**Review by vadim-kovalyov — APPROVED** [review]

File: mqtt/mqtt-bridge/src/persist/waking_state/ring_buffer/mod.rs
Comment (nyanzebra) [reply mqtt/mqtt-bridge/src/persist/waking_state/ring_buffer/mod.rs:534]: Yeah, I can double check, but I think block header size is an odd number (41 or 43?) thus cannot do offsets of 4,8, etc...

**Pull request merged**
